message/messaging_service.cc

```diff
@@ -838,9 +838,12 @@ do_make_sink_source(messaging_verb verb, uint32_t repair_meta_id, shared_ptr<mes
     return rpc_client->make_stream_sink<netw::serializer, SinkType>().then([&rpc, verb, repair_meta_id, rpc_client] (rpc::sink<SinkType> sink) mutable {
         auto rpc_handler = rpc->make_client<rpc::source<SourceType> (uint32_t, rpc::sink<SinkType>)>(verb);
         return rpc_handler(*rpc_client, repair_meta_id, sink).then_wrapped([sink, rpc_client] (future<rpc::source<SourceType>> source) mutable {
-            return (source.failed() ? sink.close() : make_ready_future<>()).then([sink = std::move(sink), source = std::move(source)] () mutable {
-                return make_ready_future<value_type>(value_type(std::move(sink), std::move(source.get0())));
-            });
+            if (source.failed()) {
+                return make_exception_future<value_type>(source.get_exception()).finally([sink = std::move(sink), rpc_client = std::move(rpc_client)] () mutable {
+                    return sink.close().finally([sink] {});
+                });
+            }
+            return make_ready_future<value_type>(value_type(std::move(sink), std::move(source.get0())));
```

Comment:
If you touch this code wouldn't it be better to co-routinize it?

Reply:
Yes, see https://github.com/scylladb/scylla/pull/10591#issuecomment-1129577691
This is tangential to this PR

Reply:
I do not see why it should be done as a follow up. This patch makes the code even more complicated and making it a co-routine will simplify it instead. Actually co-routinize it will probably solve the problem.

Reply:
I'm not sure its much simpler (if you want to return a possible sink.close() error as nested exception).
Something like this:
```c++
    using value_type = std::tuple<rpc::sink<SinkType>, rpc::source<SourceType>>;
    auto sink = co_await rpc_client->make_stream_sink<netw::serializer, SinkType>();
    auto rpc_handler = rpc->make_client<rpc::source<SourceType> (uint32_t, rpc::sink<SinkType>)>(verb);
    auto source_fut = co_await coroutine::as_future(rpc_handler(*rpc_client, repair_meta_id, sink));
    if (source_fut.failed()) {
        auto ex = source_fut.get_exception();
        try {
            co_await sink.close();
        } catch (...) {
            std::throw_with_nested(std::move(ex));
        }
        co_return coroutine::exception(std::move(ex));
    }
    co_return value_type(std::move(sink), std::move(source_fut.get0()));
```

Reply:
Your original fix does not return nested exception, no? Do not remember what finally does if body throws. But the co-routine code looks better to me.

Reply:
ok, just passed testing (unit(dev) + repair_additional_test.py::TestRepairAdditional::test_repair_kill_3(debug)), pushed v2 with coroutine.